File: solution1.py
# Time: O(n)
# Space: O(n)

class Solution(object):
    def subarraySum(self, nums, k):
        # get running sum at each index
        # traverse the running sum array and store the unique running sums and its freq in hash map
        # if a runnning sum - k is seen as a running sum n the past then increase counter by 1
        rsum = []
        rs,cnt = 0, 0
        hmap = {}
        hmap[0] = 1
        for n in nums:
            rs += n
            rsum.append(rs)
        for rs in rsum:
            if rs - k in hmap:
                cnt += hmap[rs-k]
            if rs not in hmap:
                hmap[rs] = 1
            else:
                hmap[rs] += 1
        return cnt

File: Solution1.java
import java.util.HashMap;

public class Solution1 {

    public int subarraySum(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        map.put(0, 1);
        int count = 0;
        int sum = 0;
        for (int i = 0; i < nums.length; i++) {
            sum = sum + nums[i];
            if (map.get(sum - k) != null) {
                count = count + map.get(sum - k);
            }
            map.put(sum, map.getOrDefault(sum, 0) + 1);
        }
        return count;
    }
}
